### PR TITLE
improve todo-mvc example

### PR DESCRIPTION
**Pull request by RoelVanderPaal**
- implement HTML as defined in https://github.com/tastejs/todomvc-app-template
- implement full app specification: https://github.com/tastejs/todomvc/blob/master/app-spec.md
  - toggle-all button
  - editing:
    - trim text
    - remove when empty 
    - undo on escape key
  - clear-completed button
  - localStorage Json format

### CODE REVIEW

File: todo-mvc/src/main/scala/todomvc/TodoMvc.scala [hunk 1]
```diff
@@ -124,43 +145,61 @@ object TodoMvc extends IOWebApp:
...
+              cls := "clear-completed",
+              onClick --> {
+                _.foreach(_ => store.clearCompleted)
+              },
+              "Clear completed")))
     )
 
 class TodoStore(entries: SignallingSortedMapRef[IO, Long, Todo], nextId: IO[Long]):
+  def toggleAll(completed: Boolean): IO[Unit] =
+    entries.update(_.map((id, todo) => (id, todo.copy(completed = completed))))
```
Comment (armanbilge) [inline todo-mvc/src/main/scala/todomvc/TodoMvc.scala:195]: maybe can use `mapValues` here?

File: todo-mvc/src/main/scala/todomvc/TodoMvc.scala [hunk 2]
```diff
@@ -92,22 +99,36 @@ object TodoMvc extends IOWebApp:
...
               input.withSelf { self =>
-                val endEdit = self.value.get.flatMap { text =>
-                  todo.update(_.map(_.copy(text = text))) *> editing.set(false)
-                }
-
+                val endEdit = self.value.get.map(_.trim).flatMap { text =>
+                  todo.update(t =>
+                    text match {
+                      case "" => None
+                      case _ => t.map(_.copy(text = text.trim))
```
Comment (armanbilge) [inline todo-mvc/src/main/scala/todomvc/TodoMvc.scala:106]: Is this `trim` redundant to the one above?

File: todo-mvc/src/main/scala/todomvc/TodoMvc.scala [hunk 3]
```diff
@@ -92,22 +99,36 @@ object TodoMvc extends IOWebApp:
...
+                      case _ => IO.unit
+                    }
                   },
-                  onBlur --> (_.foreach(_ => endEdit))
+                  onBlur --> (_.foreach(_ => {
+                    editing
+                      .get
+                      .flatMap(
+                        IO.whenA(_)(endEdit)
+                      ) // do not endEdit when blur is triggered after Escape
```
Comment (armanbilge) [inline todo-mvc/src/main/scala/todomvc/TodoMvc.scala:124]:
```suggestion
                    // do not endEdit when blur is triggered after Escape
                    editing.get.flatMap(IO.whenA(_)(endEdit))
```

File: todo-mvc/src/main/scala/todomvc/TodoMvc.scala [hunk 4]
```diff
@@ -124,43 +145,61 @@ object TodoMvc extends IOWebApp:
...
       span(
         cls := "todo-count",
-        activeCount.map {
-          case 1 => "1 item left"
-          case n => n.toString + " items left"
+        strong(store.activeCount.map(_.toString)),
+        store.activeCount.map {
+          case 1 => " item left"
+          case n => " items left"
         }
```
Comment (armanbilge) [inline todo-mvc/src/main/scala/todomvc/TodoMvc.scala:166]:
```suggestion
        }.changes
```

File: todo-mvc/src/main/scala/todomvc/TodoMvc.scala [hunk 5]
```diff
@@ -178,30 +214,57 @@ object TodoStore:
...
+//        .localStorage
+//        .events(window)
+//        .foreach {
+//          case Storage.Event.Updated(`key`, _, value, _) =>
+//            jawn.decode[SortedMap[Long, Todo]](value).foldMapM(mapRef.set(_))
+//          case _ => IO.unit
+//        }
+//        .compile
+//        .drain
+//        .background
```
Comment (armanbilge) [inline todo-mvc/src/main/scala/todomvc/TodoMvc.scala:257]: Why is it commented?

Comment (RoelVanderPaal) [reply todo-mvc/src/main/scala/todomvc/TodoMvc.scala]: I did not see the use case for that, it works fine without.
I uncommented it anyway.

Comment (armanbilge) [reply todo-mvc/src/main/scala/todomvc/TodoMvc.scala]: > I did not see the use case for that, it works fine without.

Try opening two tabs to the TodoMVC page at the same time :)

**Comment by RoelVanderPaal** [comment]
FYI, I tested with https://github.com/RoelVanderPaal/playwright/tree/main/examples/todomvc

**Comment by armanbilge** [comment]
> FYI, I tested with https://github.com/RoelVanderPaal/playwright/tree/main/examples/todomvc

Cool! Does Playwright have a Java API? Would be cool to have those tests integrated into our test suite. Actually, at least a couple other Scala.js UI frameworks could share them as well :)

**Review by armanbilge — APPROVED** [review]
thanks for your work on this!! looks great :)

**Comment by RoelVanderPaal** [comment]
> > FYI, I tested with https://github.com/RoelVanderPaal/playwright/tree/main/examples/todomvc
> 
> Cool! Does Playwright have a Java API? Would be cool to have those tests integrated into our test suite. Actually, at least a couple other Scala.js UI frameworks could share them as well :)

It does indeed. Porting the tests from Javascript to Java is probably not a lot of work.
Running the tests automatically is a bit more complicated since the tested application must be served over HTTP.

**Comment by armanbilge** [comment]
> Running the tests automatically is a bit more complicated since the tested application must be served over HTTP.

A bit more complicated but also doable! If you decide to port the tests from JS to Scala/Java I would love to make it part of the test suite here.

**Pull request merged**